### PR TITLE
cmd/snap-bootstrap/initramfs-mounts: mount ubuntu-data nosuid

### PR DESCRIPTION
**Pull request by anonymouse64**
This will prevent snaps from being able to ship tarballs with suid files in
them and extracting those tarballs onto writable areas like $SNAP_DATA will not
have the suid bit set/usable for those files.

Note that this is not fixing a security vulnerability per-se, but it is closing a 
loophole and is considered to be a hardening measure for UC20. 

This change will need to be made in the initramfs to apply the same 
hardening measure for UC16 and UC18.

Fixes: https://bugs.launchpad.net/stuttgart/+bug/1922083 (for UC20)

### CODE REVIEW

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 1]
```diff
@@ -1221,6 +1226,8 @@ func mountPartitionMatchingKernelDisk(dir, fallbacklabel string) error {
 		// first partition we will be mounting, we can't know if anything is
 		// corrupted yet
 		NeedsFsck: true,
+		// don't need nosuid option here, since this function is only used
+		// for ubuntu-boot and ubuntu-seed, never ubuntu-data
```
Comment (bboozzoo) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:1230]: maybe we should then rename the function to mountNonDataPartitionMathingKernelDisk?

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 2]
```diff
@@ -1424,7 +1434,11 @@ func generateMountsModeRun(mst *initramfsMountsState) error {
 
 	// TODO: do we actually need fsck if we are mounting a mapper device?
 	// probably not?
-	if err := doSystemdMount(unlockRes.FsDevice, boot.InitramfsDataDir, fsckSystemdOpts); err != nil {
+	dataMountOpts := &systemdMountOptions{
```
Comment (bboozzoo) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:1437]:
```
// fsck and mount with nosuid to prevent snaps from being able 
// to bypass the sandbox by creating suid root files  there and 
// trying to escape the sandbox
```

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 3]
```diff
@@ -1288,9 +1295,12 @@ func generateMountsCommonInstallRecover(mst *initramfsMountsState) (model *asser
 	//            mounted, we should also implement writable-paths here too as
 	//            writing it in Go instead of shellscript is desirable
 
-	// 2.3. mount "ubuntu-data" on a tmpfs
+	// 2.3. mount "ubuntu-data" on a tmpfs, and also mount with nosuid to prevent
+	// snaps from being able to bypass the sandbox by creating suid root files
+	// there and escape the sandbox
```
Comment (bboozzoo) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:1300]: s/escape/trying to escape/

AFAIU even though the process would then be running as root there's still seccomp and apparmor that should prevent further escalation, so 'trying to escape' (or circumvent like in the earlier comment)?

Comment (setharnold) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1300]: Does snapd install apparmor profiles with an attachment specification that covers all the writable areas available to each snap to write to?

Comment (anonymouse64) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1300]: Snaps when they execute always transition to their apparmor profile for the app in the snap that is being executed, but we do not have attachment specifications to cover all the writable areas that snaps can write to.

**Comment by alexmurray** [comment]
LGTM, this should provide some good additional hardening against snaps using setuid binares.

**Review by bboozzoo — APPROVED** [review]
LGTM, thanks!

**Review by mvo5 — APPROVED** [review]
Thank you

**Comment by bboozzoo** [comment]
The failures are unrelated, debian 10 failed with interfaces many test, and sid failed applying patches which needs to be addressed in master. @pedronis @mvo5 can you land it?

**Comment by pedronis** [comment]
@bboozzoo mvo marked this for 2.52 so it can't be landed immediately

**Comment by anonymouse64** [comment]
Now that 2.51 is branched I think we can land this, it would be good to land it soon to get more testing of edge with this change

**Comment by anonymouse64** [comment]
Forgot to run the nested tests with this PR, let me close and re-open this, if the nested tests are happy this can go in now
